torch/utils/_cxx_pytree.py

```diff
@@ -825,3 +838,82 @@ def __instancecheck__(self, instance: object) -> bool:
 class LeafSpec(TreeSpec, metaclass=LeafSpecMeta):
     def __new__(cls) -> "LeafSpec":
         return optree.treespec_leaf(none_is_leaf=True)  # type: ignore[return-value]
+
+
+def tree_leaves_with_path(
+    tree: PyTree,
+    is_leaf: Optional[Callable[[PyTree], bool]] = None,
+) -> List[Tuple[KeyPath, Any]]:
+    """Gets the leaves of a pytree like ``tree_leaves`` and returns each leaf's key path.
+
+    Args:
+        tree: a pytree. If it contains a custom type, that type must be
+            registered with an appropriate `tree_flatten_with_path_fn` when registered
+            with :func:`register_pytree_node`.
+        is_leaf: An extra leaf predicate function that will be called at each
+            flattening step. The function should have a single argument with signature
+            ``is_leaf(node) -> bool``. If it returns :data:`True`, the whole subtree being treated
+            as a leaf. Otherwise, the default pytree registry will be used to determine a node is a
+            leaf or not. If the function is not specified, the default pytree registry will be used.
+    Returns:
+        A list of (key path, leaf) pairs.
+    """
+    raise NotImplementedError("KeyPaths are not yet supported in cxx_pytree.")
```

Comment:
Do we have a plan for adding these? @XuehaiPan, are you OK with how the keypath API looks like? There was some discussion over at https://github.com/pytorch/pytorch/issues/113378 about performance, but IMO we shouldn't diverge from JAX unless we have a good reason to.

Reply:
happy to work with @XuehaiPan on C++ implementation. Agree that we should stay close to the JAX API.

Reply:
> Do we have a plan for adding these?
> happy to work with @XuehaiPan on C++ implementation

Yes, we should keep the Python and C++ pytrees to provide the same APIs. The current tree path implementation does not rely on the implementation detail of `tree_flatten` and `treespec`. You can just copy (or import) the key path API in Python pytree into C++ pytree and replace the `tree_flatten` with the C++ `tree_flatten`.

------

> @XuehaiPan, are you OK with how the keypath API looks like? There was some discussion over at #113378 about performance, but IMO we shouldn't diverge from JAX unless we have a good reason to.

> I agree that we should stay close to the JAX API.

If we decide to have APIs similar to JAX's pytree, I think the current implementation would be fine.

For performance, I would like to have two different path semantics, i.e., (1) path (tuple of raw indices); and (2) accessor (or typed path in #113378).

#### (1) path (tuple of raw indices)

Observation: Most nested containers are in the same type, e.g., nested `dict` or `torch.nn.Module`. The access rule to a container element is the same on every depth of the tree (`__getitem__` for `dict` and `__getattr__` for `torch.nn.Module`).

We can obtain the path entries during the tree flattening, which will only introduce a very small overhead (no auxiliary key path wrapper is applied). Because in most cases `context` and `entries` are the same.

```python
def dict_flatten(d):
    keys, values = list(d), list(d.value())
    return values, keys, keys  # children, context, entries
```

`optree` uses this approach to save the `entries` in the `treespec` object while keeping both `tree_flatten` and `tree_flatten_with_path` with good performance. This design also allows us to have only one `flatten_fn` definition in the registry. So we do not need to worry about the user-provided `flatten_fn` and `flatten_with_keys_fn` diverge.

```python
Children = Iterable[PyTree]
Context = Any
Entry = Any

flatten_fn(tree: PyTree) -> Union[
    Tuple[Children, Context],
    Tuple[Children, Context, Optional[Iterable[Entry]]]
]
```

See also: [OpTree: Registering a Container-like Custom Type as Non-leaf Nodes](https://github.com/metaopt/optree#registering-a-container-like-custom-type-as-non-leaf-nodes)

#### (2) accessor (tuple of path entries with semantics)

This is very similar to the API proposed in https://github.com/pytorch/pytorch/pull/116786#issuecomment-1888058691, I have an initial implementation for this https://github.com/XuehaiPan/optree/blob/c8002e7e984445589d7f34f12c9a03aa7f59622f/optree/registry.py#L565-L646. It is similar to [`itemgetter`](https://docs.python.org/3/library/operator.html#operator.itemgetter) and [`attrgetter`](https://docs.python.org/3/library/operator.html#operator.attrgetter) with `__call__` magic methods.

A typed path is `(entry, type, kind)`, it can be wrapped with `PyTreePathEntry`. A `PyTreeAccessor` is a tuple of `PyTreePathEntry`.

```python
class PyTreePathEntry:
    entry: Any
    type: Type
    kind: PyTreeKind

    def __call__(self, obj):
        # index into the object (based on kind or user-defined logic in a subclass)
        return obj[self.entry]


class PyTreeAccessor(tuple):  # Tuple[PyTreePathEntry, ...]
    def __call__(self, obj):
        for entry in self:
            obj = entry(obj)
        return obj
```

```python
In [1]: import optree

In [2]: from collections import *

In [3]: MyTuple = namedtuple('MyTuple', ['x', 'y', 'z'])

In [4]: tree = OrderedDict([('a', (1, 2)), ('b', {'c': [3, 4], 'd': 5}), ('e', MyTuple(6, 7, 8))])

In [5]: optree.tree_flatten_with_accessor(tree)
Out[5]: 
(
    [
        PyTreeAccessor(*['a'][0], (PyTreePathEntry(entry='a', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=0, type=<class 'tuple'>, kind=<PyTreeKind.TUPLE: 3>))),
        PyTreeAccessor(*['a'][1], (PyTreePathEntry(entry='a', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=1, type=<class 'tuple'>, kind=<PyTreeKind.TUPLE: 3>))),
        PyTreeAccessor(*['b']['c'][0], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='c', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>), PyTreePathEntry(entry=0, type=<class 'list'>, kind=<PyTreeKind.LIST: 4>))),
        PyTreeAccessor(*['b']['c'][1], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='c', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>), PyTreePathEntry(entry=1, type=<class 'list'>, kind=<PyTreeKind.LIST: 4>))),
        PyTreeAccessor(*['b']['d'], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='d', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>))),
        PyTreeAccessor(*['e'][0], (PyTreePathEntry(entry='e', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=0, type=<class '__main__.MyTuple'>, kind=<PyTreeKind.NAMEDTUPLE: 6>))),
        PyTreeAccessor(*['e'][1], (PyTreePathEntry(entry='e', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=1, type=<class '__main__.MyTuple'>, kind=<PyTreeKind.NAMEDTUPLE: 6>))),
        PyTreeAccessor(*['e'][2], (PyTreePathEntry(entry='e', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=2, type=<class '__main__.MyTuple'>, kind=<PyTreeKind.NAMEDTUPLE: 6>)))
    ],
    [1, 2, 3, 4, 5, 6, 7, 8],
    PyTreeSpec(OrderedDict([('a', (*, *)), ('b', {'c': [*, *], 'd': *}), ('e', MyTuple(x=*, y=*, z=*))]))
)

In [6]: accessors, leaves, treespec = optree.tree_flatten_with_accessor(tree)

In [7]: accessor = accessors[3]

In [8]: accessor
Out[8]: PyTreeAccessor(*['b']['c'][1], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='c', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>), PyTreePathEntry(entry=1, type=<class 'list'>, kind=<PyTreeKind.LIST: 4>)))

In [9]: accessor(tree)
Out[9]: 4

In [10]: accessor.pprint('tree')
Out[10]: "tree['b']['c'][1]"

In [11]: tree['b']['c'][1]
Out[11]: 4

In [12]: accessor[:-1](tree)
Out[12]: [3, 4]

In [13]: accessor[:-2](tree)
Out[13]: {'c': [3, 4], 'd': 5}

In [14]: accessor[0](tree)
Out[14]: {'c': [3, 4], 'd': 5}

In [15]: tree['b']
Out[15]: {'c': [3, 4], 'd': 5}
```